src/python/pants/engine/internals/native_engine.pyi

```diff
@@ -248,6 +250,97 @@ class PyExecutor:
     def to_borrowed(self) -> PyExecutor: ...
     def shutdown(self, duration_secs: float) -> None: ...
 
+# ------------------------------------------------------------------------------
+# Target
+# ------------------------------------------------------------------------------
+
+# Type alias to express the intent that the type should be immutable and hashable. There's nothing
+# to actually enforce this, outside of convention.
+ImmutableValue = Any
+
+# Marker for unspecified field values that should use the default value if applicable.
+NO_VALUE: Any
```

Comment:
It still does: https://github.com/pantsbuild/pants/pull/19143/files#diff-2448086e95c28a9e1537a6a950f161e125b5ae3b59838d4d4caf437123332c29R26 ... the type is just not exposed.

Reply:
It still does: https://github.com/pantsbuild/pants/pull/19143/files#diff-2448086e95c28a9e1537a6a950f161e125b5ae3b59838d4d4caf437123332c29R26 ... the type is just not exposed.

Reply:
Right, we should still expose these in our `.pyi` so type-checkers know the type has these (but not other methods).

E.g.
```python
print(repr(NO_VALUE))  # typesafe
NO_VALUE + 1  # Shouldn't be typesafe, but is with `Any`
```

Reply:
(unresolving and disabling auto-merge, since technically this is a regression)

Reply:
Another option for hiding the type of a sentinel like this is `object`, rather than `Any`: the latter is definitely convenient, but the "automatically any type" behaviour isn't so nice.

In my work repo, we've been frustrated with libs that set sentinels to have type `Any`, since it is so easy to forget to handle them. We generally find  explicit unions like `NoValue | None | int` work okay, but I suspect the trade off is different for pants.

Reply:
To drive @huonw's point home as well:

```python
class MyProtocol(Protocol):
    super_special_attribute: int

def my_fun(obj: MyProtocol):
    ...

sentinel1: Any
sentinel2: object

my_fun(sentinel1)  # not a type error
my_fun(sentinel2)  # a type error
```

`Any` means "literally could be anything". Such a nasty sticky type.

Reply:
Despite `NO_VALUE` not being underscore prefixed, it does not have a practical use outside of `target.py`, so I don't think that this counts as a breaking change. Nonetheless, will push an update here.